File: transport-mode/node/sendExample.js
'use strict';
// Send messages through the Roddy API — the transport-mode reply path.
//
//   npm install
//   cp .env.example .env      # fill client creds + channel/contact
//   node sendExample.js
//
// Shows: get a token (client-credentials, handled by RoddyTokenClient), send a
// text, then upload an image and send it. client_id is derived from your token.

require('dotenv').config();
const crypto = require('crypto');
const { RoddyTokenClient } = require('./auth');
const { RoddyClient, RoddySendError } = require('./roddyClient');

// A valid 1x1 PNG so the media example is self-contained.
const PIXEL_PNG = Buffer.from(
  'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg==',
  'base64'
);

async function main() {
  const client = new RoddyClient(
    process.env.RODDY_API_BASE_URL,
    new RoddyTokenClient(
      process.env.AUTH_TOKEN_URL,
      process.env.CLIENT_ID,
      process.env.CLIENT_SECRET
    )
  );
  const channelId = process.env.CHANNEL_ID;
  const contactId = process.env.CONTACT_ID;

  try {
    const t = await client.sendMessage(
      channelId, contactId,
      { message_type: 'text', text: 'Hello from the Roddy example ✅' },
      { idempotencyKey: crypto.randomUUID() }
    );
    console.log('sent text →', t.message_id);

    const mediaUrl = await client.uploadMedia(channelId, {
      mimeType: 'image/png', filename: 'pixel.png', fileBytes: PIXEL_PNG,
    });
    const i = await client.sendMessage(
      channelId, contactId,
      { message_type: 'image', media_mime_type: 'image/png', media_url: mediaUrl, text: 'an uploaded image' },
      { idempotencyKey: crypto.randomUUID() }
    );
    console.log('sent image →', i.message_id);
  } catch (e) {
    if (e instanceof RoddySendError) {
      console.error(`send failed: status=${e.status} code=${e.code} retriable=${e.retriable} — ${e.detail}`);
    } else {
      throw e;
    }
  }
}

main();

File: transport-mode/python/send_example.py
"""Send messages through the Roddy API — the transport-mode reply path.

    cp .env.example .env      # fill client creds + channel/contact
    python send_example.py

Shows: get a token (client-credentials, handled by RoddyTokenClient), send a
text, then upload an image and send it. `client_id` is derived from your token,
so it's never in the body.
"""

from __future__ import annotations

import base64
import os
import uuid

from auth import RoddyTokenClient
from roddy_client import RoddyClient, RoddySendError

# A valid 1x1 PNG, so the media example is self-contained (no asset file needed).
PIXEL_PNG = base64.b64decode(
    "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNk"
    "+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg=="
)


def main() -> None:
    client = RoddyClient(
        os.environ["RODDY_API_BASE_URL"],
        RoddyTokenClient(
            os.environ["AUTH_TOKEN_URL"],
            os.environ["CLIENT_ID"],
            os.environ["CLIENT_SECRET"],
        ),
    )
    channel_id = os.environ["CHANNEL_ID"]
    contact_id = os.environ["CONTACT_ID"]

    try:
        # 1) A plain text message.
        r = client.send_message(
            channel_id,
            contact_id,
            {"message_type": "text", "text": "Hello from the Roddy example ✅"},
            idempotency_key=str(uuid.uuid4()),
        )
        print("sent text →", r.get("message_id"))

        # 2) Media: upload the bytes, then send by the returned (owned) URL.
        media_url = client.upload_media(
            channel_id, mime_type="image/png", filename="pixel.png", file_bytes=PIXEL_PNG
        )
        r = client.send_message(
            channel_id,
            contact_id,
            {
                "message_type": "image",
                "media_mime_type": "image/png",
                "media_url": media_url,
                "text": "an uploaded image",
            },
            idempotency_key=str(uuid.uuid4()),
        )
        print("sent image →", r.get("message_id"))

    except RoddySendError as e:
        # Branch on the stable code; retry only when retriable.
        print(f"send failed: status={e.status} code={e.code} retriable={e.retriable} — {e.message}")


if __name__ == "__main__":
    main()
